File: Nikhil/python/scenario_component.py
"""
scenario_component.py

Python translation of `com.infosys.small.pnbc.ScenarioComponent`.

The class loads scenario definitions from a folder containing one or more
*.json* files.  Each file must contain a JSON array whose elements conform
to the schema defined by :class:`Scenario`.

The implementation mirrors the original Java semantics while embracing
Pythonic conventions and type hints.
"""

from __future__ import annotations

import json
import logging
import os
from pathlib import Path
from typing import Any, Dict, List, Mapping, Optional

from pydantic import BaseModel, Field, ValidationError

# --------------------------------------------------------------------------- #
# Logging configuration (equivalent to Java SimpleLogger)
# --------------------------------------------------------------------------- #
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(name)s [%(levelname)s] %(message)s",
)
logger = logging.getLogger(__name__)


# --------------------------------------------------------------------------- #
# Pydantic models mirroring the nested Java static classes
# --------------------------------------------------------------------------- #
class Output(BaseModel):
    type: str
    value: str


class ScenarioToolCall(BaseModel):
    tool_id: str = Field(alias="tool_id")
    input: Dict[str, Any]
    output: List[Output]

    model_config = {"populate_by_name": True}


class Scenario(BaseModel):
    id: str
    description: str
    success_criteria: str = Field(alias="success_criteria")
    tool_calls: List[ScenarioToolCall] = Field(alias="tool_calls")

    model_config = {"populate_by_name": True}


class ScenarioMetadata(BaseModel):
    id: str = Field(alias="id")
    description: str = Field(alias="description")

    model_config = {"populate_by_name": True}


# --------------------------------------------------------------------------- #
# Main component
# --------------------------------------------------------------------------- #
class ScenarioComponent:
    """
    Access-point for scenario definitions loaded from disk.

    Parameters
    ----------
    folder : str | os.PathLike
        Directory containing one or more JSON files with scenario definitions.
    """

    # ------------------------- construction ---------------------------- #
    def __init__(self, folder: str | os.PathLike) -> None:
        if folder is None:
            raise ValueError("folder must not be None")

        self.scenario_folder: Path = Path(folder)
        if not self.scenario_folder.is_dir():
            raise IOError(f"Scenario folder {self.scenario_folder} does not exist.")

        self._scenarios: List[Scenario] = []
        self._load_scenarios()

    # Factory mirroring Java getInstance() ----------------------------- #
    @classmethod
    def get_instance(cls) -> "ScenarioComponent":
        default = Path(
            "D:/Users/mzatt/Projects/DELETEME PnBC/pnbc-services/src/main/resources/scenarios"
        )
        return cls(default)

    # --------------------------- API ---------------------------------- #
    def list_scenarios(self) -> List[Scenario]:
        """Return a *copy* of all scenarios loaded in memory."""
        return list(self._scenarios)

    def get_scenario(self, scenario_id: str) -> Optional[Scenario]:
        """
        Return the scenario with *scenario_id*, or *None* if not found.
        """
        if scenario_id is None:
            raise ValueError("scenario_id must not be None")

        for s in self._scenarios:
            if scenario_id == s.id:
                return s
        return None

    def get_success_criteria(self, scenario_id: str) -> Optional[str]:
        """Return the success criteria for *scenario_id*, if available."""
        scenario = self.get_scenario(scenario_id)
        return None if scenario is None else scenario.success_criteria

    def get(
        self,
        scenario_id: str,
        tool_id: str,
        args: Mapping[str, Any],
    ) -> str:
        """
        Retrieve the **text** output for the specified tool invocation.

        The logic follows the original Java implementation:
        * If the scenario does not exist → return an *ERROR* string.
        * Iterate through tool calls, match both *tool_id* and *args*.
        * Concatenate the value of every output whose type is ``"text"``.
        * If no matching call is found → generic *ERROR* string.
        """
        if scenario_id is None or tool_id is None or args is None:
            raise ValueError("scenario_id, tool_id and args must not be None")

        scenario = self.get_scenario(scenario_id)
        if scenario is None:
            return f"ERROR: Scenario {scenario_id} does not exist."

        for call in scenario.tool_calls:
            if call.tool_id != tool_id:
                continue
            if not self._matched(call.input, args):
                continue
            return "".join(o.value for o in call.output if o.type == "text")

        return "ERROR: System failure, wrong API call parameters."

    # --------------------- internal helpers --------------------------- #
    def _load_scenarios(self) -> None:
        """Populate :pyattr:`_scenarios` from JSON files in *scenario_folder*."""
        files = [p for p in self.scenario_folder.iterdir() if p.suffix == ".json"]
        if not files:
            raise IOError("Scenario folder is empty or contains no *.json* files.")

        for file in files:
            try:
                with file.open(encoding="utf-8") as fh:
                    data = json.load(fh)
                self._scenarios.extend(Scenario.model_validate(o) for o in data)
            except (json.JSONDecodeError, ValidationError) as exc:
                logger.error("Error parsing scenario %s", file.name, exc_info=exc)
                raise

    # Map-normalisation & matching logic (verbatim port) --------------- #
    @staticmethod
    def _transform_map(raw: Mapping[str, Any]) -> Dict[str, str]:
        result: Dict[str, str] = {}
        for k, v in raw.items():
            new_key = "" if k is None else str(k).lower().strip()
            new_val = "" if v is None else str(v).lower().strip()
            result[new_key] = new_val
        return result

    @classmethod
    def _matched(cls, map1: Mapping[str, Any], map2: Mapping[str, Any]) -> bool:
        t1 = cls._transform_map(map1)
        t2 = cls._transform_map(map2)
        for k, v in t1.items():
            if k not in t2 or t2[k] != v:
                return False
        return True

File: Nikhil/java/src/main/java/com/infosys/small/pnbc/ScenarioComponent.java
package com.infosys.small.pnbc;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.NonNull;

/**
 * This is used to access scenarios.
 */
public class ScenarioComponent {

	private final static Logger LOG = LoggerFactory.getLogger(ScenarioComponent.class);

	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	public static class Scenario {
		private String id;
		private String description;

		@JsonProperty("success_criteria")
		private String successCriteria;

		@JsonProperty("tool_calls")
		private List<ScenarioToolCall> toolCalls;
	}

	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	public static class ScenarioToolCall {
		@JsonProperty("tool_id")
		private String toolId;

		private Map<String, Object> input;
		private List<Output> output;
	}

	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	public static class Output {
		private String type;
		private String value;
	}

	@Data
	@AllArgsConstructor
	public static class ScenarioMetadata {
		@JsonProperty("id")
		private String id;

		@JsonProperty("description")
		private String description;
	}

	private final File scenarioFolder;
	private final ObjectMapper objectMapper = new ObjectMapper();
	private final List<Scenario> scenarios = new ArrayList<>();

	public static ScenarioComponent getInstance() throws IOException {
		return new ScenarioComponent(
				new File("D:\\Users\\mzatt\\Projects\\DELETEME PnBC\\pnbc-services\\src\\main\\resources\\scenarios"));
	}

	public ScenarioComponent(@NonNull File folder) throws IOException {
		this.scenarioFolder = folder;

		File[] files = scenarioFolder.listFiles((dir, name) -> name.endsWith(".json"));
		if (files == null)
			throw new IOException("Scenario folder is empty or inaccessible");

		for (File file : files) {
			try {
				scenarios.addAll(objectMapper.readValue(file, new TypeReference<List<Scenario>>() {
				}));
			} catch (IOException e) {
				LOG.error("Error parsing scenario " + file.getName(), e);
				throw e;
			}
		}
	}

	public List<Scenario> listScenarios() {
		return new ArrayList<>(scenarios);
	}

	/**
	 * 
	 * @param scenarioId
	 * @return Given scenario, or null if it cannot be found.
	 */
	public Scenario getScenario(@NonNull String scenarioId) throws IOException {

		for (Scenario s : listScenarios())
			if (scenarioId.equals(s.getId())) {
				return s;
			}
		return null;
	}

	public String getSuccessCriteria(@NonNull String scenarioId) {

		Scenario scenario = null;
		try {
			scenario = getScenario(scenarioId);
		} catch (IOException e) {
			return null;
		}

		if (scenario == null)
			return null;
		return scenario.getSuccessCriteria();
	}

	public String get(@NonNull String scenarioId, @NonNull String toolId, @NonNull Map<String, Object> args)
			throws IOException {

		Scenario scenario = getScenario(scenarioId);

		if (scenario == null)
			return "ERROR: Scenario " + scenarioId + " does not exist.";

		for (ScenarioToolCall call : scenario.getToolCalls()) {
			if (!toolId.equals(call.getToolId()))
				continue;

			if (!matched(call.getInput(), args))
				continue;

			return call.getOutput().stream().filter(output -> "text".equals(output.getType())).map(Output::getValue)
					.reduce("", (a, b) -> a + b);
		}

		return "ERROR: System failure, wrong API call parameters.";
	}

	private static Map<String, String> transformMap(Map<String, Object> input) {
		Map<String, String> result = new java.util.HashMap<>();
		for (Map.Entry<String, Object> entry : input.entrySet()) {
			String newKey = entry.getKey() == null ? "" : entry.getKey().toLowerCase().trim();
			String newValue = entry.getValue() == null ? "" : entry.getValue().toString().toLowerCase().trim();
			result.put(newKey, newValue);
		}
		return result;
	}

	private static boolean matched(Map<String, Object> map1, Map<String, Object> map2) {
		Map<String, String> t1 = transformMap(map1);
		Map<String, String> t2 = transformMap(map2);
		for (Map.Entry<String, String> entry : t1.entrySet()) {
			if (!t2.containsKey(entry.getKey()) || !t2.get(entry.getKey()).equals(entry.getValue())) {
				return false;
			}
		}
		return true;
	}

	public static void main(String args[]) throws IOException {
		ScenarioComponent instance = ScenarioComponent.getInstance();
		System.out.println(instance.get("scenario-01", "getUnassignedTasks", new HashMap<>()));
	}
}
